notify/lark/lark.go

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright (c) 2022, MegaEase
+ * All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package lark
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/megaease/easeprobe/global"
+	"github.com/megaease/easeprobe/notify/base"
+	"github.com/megaease/easeprobe/report"
+	log "github.com/sirupsen/logrus"
+)
+
+// NotifyConfig is the slack notification configuration
+type NotifyConfig struct {
+	base.DefaultNotify `yaml:",inline"`
+	WebhookURL         string `yaml:"webhook"`
+}
+
+// Kind return the type of Notify
+func (c *NotifyConfig) Kind() string {
+	return c.MyKind
+}
+
+// Config configures the slack notification
+func (c *NotifyConfig) Config(gConf global.NotifySettings) error {
+	c.MyKind = "lark"
+	c.Format = report.Lark
+	c.SendFunc = c.SendLark
+	c.DefaultNotify.Config(gConf)
+	log.Debugf("Notification [%s] - [%s] configuration: %+v", c.MyKind, c.Name, c)
+	return nil
+}
+
+// SendLark is the wrapper for SendLarkNotification
+func (c *NotifyConfig) SendLark(title, msg string) error {
+	return c.SendLarkNotification(msg)
+}
+
+// SendLarkNotification will post to an 'Robot Webhook' url in Lark Apps. It accepts
+// some text and the Lark robot will send it in group.
+func (c *NotifyConfig) SendLarkNotification(msg string) error {
+	req, err := http.NewRequest(http.MethodPost, c.WebhookURL, bytes.NewBuffer([]byte(msg)))
+	if err != nil {
+		return err
+	}
+
+	req.Header.Add("Content-Type", "application/json")
+	req.Close = true
+
+	client := &http.Client{Timeout: c.Timeout}
+	resp, err := client.Do(req)
+	if err != nil {
+		return err
+	}
+	if resp.Body != nil {
+		defer resp.Body.Close()
+	}
+
+	buf, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return err
+	}
+	ret := make(map[string]interface{})
+	err = json.Unmarshal(buf, &ret)
+	if err != nil {
+		return fmt.Errorf("Error response from Lark [%d] - [%s]", resp.StatusCode, string(buf))
+	}
+	// Server returns {"Extra":null,"StatusCode":0,"StatusMessage":"success"} on success
+	// otherwise it returns {"code":9499,"msg":"Bad Request","data":{}}
+	if ret["StatusCode"] != "0" {
+		return fmt.Errorf("Error response from Lark [%d] - [%s]", int(ret["code"].(float64)), ret["msg"])
+	}
```

Comment:
I mean the HTTP Status Code.  it's  `resp.StatusCode` - the status code for HTTP instead of Lark API's. for examples,  200 means success,  302 means redirect, 403 means permission, 503 means system error.

So, doesn't the Lark use this way to return the success?

Reply:
When success, server returns HTTP status code `200` with HTTP response body `{"Extra":null,"StatusCode":0,"StatusMessage":"success"}`, so if the `StatusCode` in response body not equals to `"0"`, it's must be a error response, and we can use the `code` field to indicate the error.

So, I use the code in response body to instead of the HTTP status code.

Reply:
OK, If you confirm this, I'm going to approve this PR with this uncommon and silly design. :joy: